### DISCUSSION TITLE
How to validation input type="file"

### DISCUSSION THREAD

**Discussion by NantipatSoftEn** [react-hook-form/react-hook-form#1946] [Q&A] [answered]
I read this document https://react-hook-form.com/get-started#Applyvalidation not see example validate file upload

For example file name , max file size, type file 

 don't have this feature ? or something I missed   😷😷

**Comment by bluebill1049** [comment, marked as answer]
U will have to use validate function to measure the size of ur file, for required field required still works.

**Reply by NantipatSoftEn** [reply]
OK , Thank

**Reply by surajpheudin** [reply]
Hey can you comment the code of validating size of file?

**Reply by SalahAdDin** [reply]
It is not hard actually, you will require to get the file and not only its path and validate it with a custom schema.

**Comment by tuyenphk** [comment]
How can I add the validation for image size, image type, and display alert message if it is not suitable?
My code is here
```
const Upload = () => {
  let [file, setFile] = useState(null);
  /**
   * Create a new state variable to hold the name of the
   * file. This is needed since the browser does not retain
   * the filename by default when uploading. 
   */
  let [filename, setFilename] = useState('');

  const filterBySize = (file) => {
    //filter out files larger than 5MB
    return file.size < 5242880;
  };

  const fileChangedHandler = event => {
    let file = event.target.files[0];
    let reader = new FileReader();
    /**Capture filename */
    setFilename(file.name);
    console.log(file);
    reader.onload = function(e) {

      setFile(e.target.result);
    };
    reader.readAsDataURL(event.target.files[0]);

    if (file != ".png" || file !=".jpg" ) {
      window.alert("File does not support. You must use .png or .jpg ");
      return false;
   }
   if (file.size > e6) {
     window.alert("Please upload a file smaller than 1 MB");
     return false;
   }
   
  };
return(
  <input className="btn btn-secondary" 
                  id="fileInput" 
                  type="file" 
                  accept="image/*"
                  // inputProps={{ accept: 'image/*' }}
                  // fileFilter={filterBySize}
                  onChange={fileChangedHandler} 
                />
                

```

**Reply by ognitio-technologies** [reply]
Try this code, this will work, InputField is material ui component you can use input instead of it

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();

 
  const submitHandler = (data) => {
   console.log(data)
  };

 <form onSubmit={handleSubmit(submitHandler)}>
<InputField
              id="image"
              label="Image"
              type="file"
              required
              inputProps={{
                ...register("image", {
                  required: "This field is required",
                  validate: {
                    lessThan1MB: (files) =>
                      files[0]?.size < 1000000 || "Max 1MB Allowed",
                    acceptedFormats: (files) =>
                      ["image/jpeg", "image/jpg"].includes(files[0]?.type) ||
                      "Only (JPG, JPEG) are allowed",
                  },
                }),
              }}
              error={errors.image && true}
              helperText={errors.image && errors.image.message}
            />
<button>submit</button>
</form>

**Comment by Dathin** [comment]
If you are wondering how to do this using just react hook form you can use the validate function.
![image](https://user-images.githubusercontent.com/34101091/125865300-507deb1a-2f93-4731-a847-30603d683373.png)
You can find more examples at the [docs](https://react-hook-form.com/api/useform/register/).

Here is my implementation of this validation:
```ts
          {...register('image', {
            validate: {
              lessThan10MB: files => files[0]?.size < 10000000 || 'Max 10MB',
              acceptedFormats: files =>
                ['image/jpeg', 'image/png', 'image/gif'].includes(
                  files[0]?.type
                ) || 'Only PNG, JPEG e GIF',
            },
          })}
```
And if you are wondering whats inside files this is the answer so you can do name validation and stuff
![image](https://user-images.githubusercontent.com/34101091/125880033-400ad0e6-b750-4725-91f5-d2d94462e349.png)

**Reply by Boye-dev** [reply]
It doesnt work

**Reply by Ali-Aref** [reply]
can't access the `files[0].type` or `files[0].size`, however I can access the file as `file` without array, and `file.type` and `file.size` returns `undefined`

**Reply by saharbca** [reply]
> can't access the `files[0].type` or `files[0].size`, however I can access the file as `file` without array, and `file.type` and `file.size` returns `undefined`

Did get a solution for that ?

**Reply by vqtruong0910** [reply]
you should use setValue("image", {e.target.files[0]}) in onChange event of input tag before using above code

**Comment by rootsher** [comment]
https://github.com/react-hook-form/react-hook-form/issues/127#issuecomment-963884695

**Comment by SalahAdDin** [comment]
There is a problem with this approach: You need to use a custom handle function to process from the file path to a file object itself, but doing that you will need to leave the input value out to avoid [this problem](https://stackoverflow.com/questions/70730587/getting-invalidstateerror-an-attempt-was-made-to-use-an-object-that-is-not-or).

But by avoiding passing the value to the controlled input, we can not reset it using the `reset` function.

That's a big problem when we have a static form, for instance.

**Comment by muzudre** [comment]
This worked for me:
```js
ref={register({
    required: 'File is required',
    validate: (value) => {
        const acceptedFormats = ['pdf']; // Add more file formats if needed
        const fileExtension = value[0]?.name.split('.').pop().toLowerCase();
        if (!acceptedFormats.includes(fileExtension)) {
            return 'Invalid file format. Only PDF files are allowed.';
        }
        return true;
    }
})}
```
